packages/grafana-prometheus/src/querybuilder/shared/interfaces.ts

```diff
@@ -0,0 +1,18 @@
+import { PromQueryModeller } from '../PromQueryModeller';
+import { PromQueryPattern } from '../types';
+
+import { LokiAndPromQueryModellerBase } from './LokiAndPromQueryModellerBase';
+
+export interface QueryModeller extends LokiAndPromQueryModellerBase {
+  getQueryPatterns(): PromQueryPattern[];
+}
+
+// Create a singleton registry
+let modellerInstance: QueryModeller | null = null;
+
+export function getModeller(): QueryModeller {
+  if (!modellerInstance) {
+    modellerInstance = new PromQueryModeller();
+  }
+  return modellerInstance;
+}
```

Comment:
Could you please provide some comment about why we need a singleton here?

Reply:
I discovered there was one last pesky circular dependency. This was fixed in https://github.com/grafana/grafana/pull/105575/commits/0905eafa8cd959cd95979391d6cebbddc4077588, and the pattern changed a bit. In that commit, I added comments to describe why we use some special patterns (to avoid circular deps).